src/resource/resource.c

```diff
@@ -471,19 +494,28 @@ int sid_resource_create_signal_event_source(sid_resource_t *res, sid_resource_ev
 		goto fail;
 	}
 
-	/* sd_event_add_signal requires the signal to be blocked before! */
-	if ((r = sigemptyset(&ss)) < 0 ||
-	    (r = sigaddset(&ss, signal)) < 0 ||
-	    (r = sigprocmask(SIG_BLOCK, &ss, NULL) < 0))
+	if (sigprocmask(SIG_BLOCK, &mask, NULL) < 0) {
+		log_error(ID(res), "Failed to set sigprocmask().");
+		goto fail;
+	}
+
+	res_event_loop->event_loop.signalfd = signalfd (-1, &mask, 0);
+	if (res_event_loop->event_loop.signalfd < 0) {
+		log_error(ID(res), "Failed to create signalfd.");
 		goto fail;
+	}
 
-	if ((r = sd_event_add_signal(res_event_loop->sd_event_loop, &sd_es, signal, handler ? _sd_signal_event_handler : NULL, NULL)) < 0)
+	if ((r = _create_event_source(res, name, sd_es, handler, NULL, es)) < 0)
```

Comment:
...the `_create_event_source` call should be after the `sd_event_add_io` because the `_create_event_source` creates the wrapper around the `sd_es` that is returned from `sd_event_add_io`.

The `_create_event_source` then calls `sd_event_source_set_userdata(sd_es, new_es)` where the `new_es` is our wrapper which then makes it possible to access this in the handler by calling  `sd_event_get_userdata` easily.

So then it would look like:

```
if  (sd_event_add_io(res_event_loop->event_loop.sd_event_loop, &sd_es, res_event_loop->event_loop.signalfd, EPOLLIN, _sd_signal_event_handler, NULL) < 0) {
 ...
}

if ((r = _create_event_source(res, name, sd_es, handler, data)) < 0) {
...
}
```

However, then there's one issue I can think of here - if we wanted one signalfd per resource AND we want to support more than one `sid_resource_create_signal_event_source` call per resource with possible different `handler` and `data` passed in, then this comes short - we'd call the `sd_event_add_io` with `_create_event_source` only once and not more. So this needs some refinement on how to deal with this (...but systemd itself deals with it because it already uses only one signalfd even though we may call `sd_event_add_signal` several times). So we should probably do something similar to what systemd does (I haven't looked yet).

Reply:
Just peeked at systemd code: https://github.com/systemd/systemd/blob/79dbbb261d245b619a4d82c1cdf96f6a4310ee0d/src/libsystemd/sd-event/sd-event.c#L1345-L1354

So, yes, they're simply storing an array indexed by signal number. Then each item in the array can hold the `callback` + `userdata` (`handler` + `data` in SID terms). This way, they can add and remove signal handlers on demand while still using the same and single `signalfd` (signalfd can be still altered after creating - just passing a new signal mask each time we alter - when we call `signalfd` with the first arg being `NULL`, we create the FD with the signal mask... then if we pass that FD to any subsequent `signalfd` call, we alter the mask).